### PR TITLE
🐛 Specify BIOS and instance UUIDs when creating VM

### PR DESCRIPTION
**Pull request by dougm**
**What does this PR do, and why is it needed?**

When creating a new VirtualMachine:
- InstanceUUID is set to VM ObjectMeta.UID
- BiosUUID is included in VirtualMachineSpec
- cloud-init metadata instance ID is set to VM spec.biosUUID

Enable find by UUID preference in vcenter.GetVirtualMachine, as it is more stable than Inventory path.

### CODE REVIEW

**Review by akutz — COMMENTED** [review]
Great start! Left some comments. Also, PR description mentions cloud-init, but I don't see that change in this PR?

File: pkg/vmprovider/providers/vsphere2/virtualmachine/configspec.go
Comment (dougm) [reply pkg/vmprovider/providers/vsphere2/virtualmachine/configspec.go]: Good catch, yes. Fixed and added test.

File: api/v1alpha2/virtualmachine_types.go
Comment (dougm) [reply api/v1alpha2/virtualmachine_types.go]: Thank you, done

File: webhooks/virtualmachine/v1alpha2/mutation/virtualmachine_mutator.go
Comment (bryanv) [inline webhooks/virtualmachine/v1alpha2/mutation/virtualmachine_mutator.go:316]: Nit: we could move this inside the `if id != vm.Spec.BiosUUID` block since that's when we need to write to the map.

File: pkg/vmprovider/providers/vsphere2/virtualmachine/configspec_test.go
Comment (bryanv) [inline pkg/vmprovider/providers/vsphere2/virtualmachine/configspec_test.go:9]: Nit: sort this down below with the rest of the non dot imports (we should prb see if there's a linter for this).

File: api/v1alpha2/virtualmachine_types.go
Comment (bryanv) [inline api/v1alpha2/virtualmachine_types.go:389]: Can you set this field in the big https://github.com/vmware-tanzu/vm-operator/blob/main/api/v1alpha1/virtualmachine_conversion_test.go#L61 conversion test to make sure that its preserved on conversions?

File: pkg/vmprovider/providers/vsphere2/vcenter/getvm.go
Comment (bryanv) [inline pkg/vmprovider/providers/vsphere2/vcenter/getvm.go:34]: Do we want to do this lookup first since it is support to be the most stable?

Should we only do this lookup when this UID matches the Spec value?

File: api/v1alpha2/virtualmachine_types.go
Comment (dougm) [reply api/v1alpha2/virtualmachine_types.go:389]: Thanks for pointing that out, done.

File: pkg/vmprovider/providers/vsphere2/vcenter/getvm.go
Comment (dougm) [reply pkg/vmprovider/providers/vsphere2/vcenter/getvm.go]: Yes, changed the order.

For new VMs:
- `InstanceUuid` is set to the VM's `ObjectMeta.UID`
- `Uuid` is set to VM's spec.biosUUID

For existing brownfield VMs, both InstanceUuid and Uuid were generated by VC. So I don't think we have a way to skip? But, we could add lookup for BiosUUID, as in:
```
findVMByUUID(vmCtx, vimClient, datacenter, vmCtx.VM.Spec.BiosUUID, false)
```
What do you think?

File: pkg/vmprovider/providers/vsphere2/virtualmachine/configspec_test.go
Comment (dougm) [reply pkg/vmprovider/providers/vsphere2/virtualmachine/configspec_test.go]: Done

File: webhooks/virtualmachine/v1alpha2/mutation/virtualmachine_mutator.go
Comment (dougm) [reply webhooks/virtualmachine/v1alpha2/mutation/virtualmachine_mutator.go]: Done

**Review by dougm — COMMENTED** [review]
Thanks for the review @bryanv

**Comment by dougm** [comment]
> Great start! Left some comments. Also, PR description mentions cloud-init, but I don't see that change in this PR?

@akutz sorry just realized I'd missed this comment. The PR propagates spec.biosUUID to cloud-init UID via `InstanceIDAnnotation`: https://github.com/vmware-tanzu/vm-operator/pull/463/files#diff-c8fd712c0c230b622f8d05fad6eb15f695c439d43330d4c863ca6316543f6c22R336

**Comment by akutz** [comment]
Hey @dougm ,

You will need to add code to https://github.com/vmware-tanzu/vm-operator/blob/5b7945822399a295d3d4e5c792e1dae667cb710d/api/v1alpha1/virtualmachine_conversion.go#L1060-L1083 and  https://github.com/vmware-tanzu/vm-operator/blob/5b7945822399a295d3d4e5c792e1dae667cb710d/api/v1alpha2/virtualmachine_conversion.go#L13-L33 to restore the value of `spec.biosUUID` when converting from v1a1 or v1a2 back to v1a3.

**Comment by dougm** [comment]
> to restore the value of `spec.biosUUID` when converting from v1a1 or v1a2 back to v1a3.
Done, thanks @akutz

File: api/v1alpha1/virtualmachine_conversion.go
Comment (bryanv) [inline api/v1alpha1/virtualmachine_conversion.go:1081]: To keep the same pattern can we have a `restore_v1alpha3_VirtualMachineBiosUUID()` function like we do above?

Comment (dougm) [reply api/v1alpha1/virtualmachine_conversion.go]: Done 👍

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [inline api/v1alpha3/virtualmachine_types.go:415]: This should also validate the format, ex.

```go
// +kubebuilder:validation:Format:=uuid4
```

See https://github.com/kubernetes/kubernetes/blob/ac8ac0fc170eec7cdf292d841f8d026a65ce1681/vendor/github.com/go-openapi/strfmt/format.go#L96-L169 for a list of not super well-known formats that can be used.

File: webhooks/virtualmachine/mutation/virtualmachine_mutator.go
Comment (akutz) [inline webhooks/virtualmachine/mutation/virtualmachine_mutator.go:326]: Can we make this New4 or New5? Just something that makes the version of the UUID explicit. That way we can ensure the validation is consistent with the expected version.

File: api/v1alpha3/virtualmachine_types.go
Comment (dougm) [reply api/v1alpha3/virtualmachine_types.go:415]: Nice, will do.

File: webhooks/virtualmachine/mutation/virtualmachine_mutator.go
Comment (dougm) [reply webhooks/virtualmachine/mutation/virtualmachine_mutator.go:326]: I don't see those in the [uuid package](https://pkg.go.dev/github.com/google/uuid), just New, New6, New7. Should I add a helper/util func in vm-operator instead?

File: webhooks/virtualmachine/mutation/virtualmachine_mutator_intg_test.go
Comment (sreyasn) [inline webhooks/virtualmachine/mutation/virtualmachine_mutator_intg_test.go:124]: typo: 'to PoweredOn'

File: webhooks/virtualmachine/mutation/virtualmachine_mutator_intg_test.go
Comment (sreyasn) [inline webhooks/virtualmachine/mutation/virtualmachine_mutator_intg_test.go:141]: let's just call this 'vm' instead of 'modified' because it doesn't get mutated?

File: webhooks/virtualmachine/mutation/virtualmachine_mutator_intg_test.go
Comment (sreyasn) [inline webhooks/virtualmachine/mutation/virtualmachine_mutator_intg_test.go:157]: same nit as above

File: webhooks/virtualmachine/mutation/virtualmachine_mutator_unit_test.go
Comment (sreyasn) [inline webhooks/virtualmachine/mutation/virtualmachine_mutator_unit_test.go:282]: is 'to PoweredOn' a typo?

File: webhooks/virtualmachine/mutation/virtualmachine_mutator_intg_test.go
Comment (dougm) [reply webhooks/virtualmachine/mutation/virtualmachine_mutator_intg_test.go]: Whoops, fixed. I'd copy-n-pasted `It` block from another test

File: webhooks/virtualmachine/mutation/virtualmachine_mutator.go
Comment (akutz) [reply webhooks/virtualmachine/mutation/virtualmachine_mutator.go:326]: Huh. Well, do we know what version of the UUID vSphere uses?

Comment (akutz) [reply webhooks/virtualmachine/mutation/virtualmachine_mutator.go:326]: FWIW, the checks in kube are just regexes https://github.com/kubernetes/kubernetes/blob/ac8ac0fc170eec7cdf292d841f8d026a65ce1681/vendor/github.com/go-openapi/strfmt/default.go#L61-L79

Comment (akutz) [reply webhooks/virtualmachine/mutation/virtualmachine_mutator.go:326]: According to @mikekim, VMware tends to prefer UUID4, actual "random" UUIDs. The Google UUID package lists the various types and what they are -- https://pkg.go.dev/github.com/google/uuid#New. It looks like `New` creates a UUID4, so we should be good with that.

Comment (dougm) [reply webhooks/virtualmachine/mutation/virtualmachine_mutator.go:328]: Agreed, added comments indicating we're using random UUID4 here.

File: webhooks/virtualmachine/validation/virtualmachine_validator.go
Comment (dilyar85) [inline webhooks/virtualmachine/validation/virtualmachine_validator.go:965]: A couple of questions about validating the `vm.spec.biosUUID` field:
- Will we add it for old VMs during the upgrade process? If so, this check would always fail the upgrade requests.
- Do we allow DevOps users to create a VM with a specified biosUUID? I don't see it mentioned in the API or enforced in the validation webhook, but I remember that being part of your design.

File: api/v1alpha3/virtualmachine_types.go
Comment (dilyar85) [inline api/v1alpha3/virtualmachine_types.go:416]: I think this might break the B/R workflow for VMs using the cloud-init bootstrap provider. A new cloud-init instance ID is always required to ensure that the network configuration runs on the restored VM. If it's set from the `vm.spec.biosUUID`, the same value from the backup would be used in restored VMs and hence not trigger the network configuration.

Comment (dougm) [reply api/v1alpha3/virtualmachine_types.go:416]: > A new cloud-init instance ID is always required

Existing code comment says same ID should be used?
https://github.com/vmware-tanzu/vm-operator/blob/df36cc7c94996357e5042cc1ff0d0feb6a818f9f/pkg/providers/vsphere/vmlifecycle/bootstrap_cloudinit.go#L53-L54

But I've yet to read up on details of the workflow. Where else would `InstanceIDAnnotation` be set and to what value?

Comment (dilyar85) [reply api/v1alpha3/virtualmachine_types.go:416]: That comment about B/R is no longer applicable and should be removed. We explicitly delete this annotation during the restore to ensure a new cloud-init instance ID is set to the restored VM CR UID.  
The annotation is actually set during the upgrade process to the value of `iid-datasource-none`. This is to address an old bug when cloud-init was used sans userdata. See https://github.com/vmware-tanzu/vm-operator/pull/211 that initiated added this code in v1a1.

File: webhooks/virtualmachine/mutation/virtualmachine_mutator.go
Comment (aruneshpa) [inline webhooks/virtualmachine/mutation/virtualmachine_mutator.go:332]: Would this also handle all the cases and be a bit more readable?
```go
if vm.Annotations == nil {
    vm.Annotations = map[string]string{}
}
                                                                                                                                                                                                                                                    	if vm.Spec.Bootstrap != nil && vm.Spec.Boot
if vm.Annotations[vmopv1.InstanceIDAnnotation] != vm.Spec.BiosUUID {
    vm.Annotations[vmopv1.InstanceIDAnnotation] = vm.Spec.BiosUUID
    wasMutated = true
}

```

Comment (dougm) [reply webhooks/virtualmachine/mutation/virtualmachine_mutator.go:332]: I'd meant to only create the map if nil, but realized after your comment I missed checking for nil map above:
```go
id := vm.Annotations[vmopv1.InstanceIDAnnotation]
```
This code may also go away after we resolve the cloud-init.UID discussion

File: webhooks/virtualmachine/validation/virtualmachine_validator.go
Comment (dougm) [reply webhooks/virtualmachine/validation/virtualmachine_validator.go]: Thanks @dilyar85 , both should be fixed now

File: api/v1alpha3/virtualmachine_types.go
Comment (dougm) [reply api/v1alpha3/virtualmachine_types.go:444]: Updated to use new `VirtualMachineBootstrapCloudInitSpec.InstanceID` field

File: pkg/providers/vsphere/vmlifecycle/bootstrap_cloudinit.go
Comment (dilyar85) [inline pkg/providers/vsphere/vmlifecycle/bootstrap_cloudinit.go:55]: I suppose this is for backward compatibility to avoid updating an existing VM's cloud-init instance ID, right? It would be great if you could add a comment for clarity and future reference.

File: webhooks/virtualmachine/validation/virtualmachine_validator.go
Comment (dougm) [reply webhooks/virtualmachine/validation/virtualmachine_validator.go:1174]: `validateBiosUUID` is only called from `ValidateCreate`, not used in `ValidateUpdate`. I added a comment.

File: pkg/providers/vsphere/vmlifecycle/bootstrap_cloudinit.go
Comment (dougm) [reply pkg/providers/vsphere/vmlifecycle/bootstrap_cloudinit.go]: Right, added comment.

**Review by dilyar85 — APPROVED** [review]
LGTM! Once this change is merged, we need to update the restore codebase accordingly to ensure a new cloud-init instance ID is used for creating a restored VM.

File: webhooks/virtualmachine/mutation/virtualmachine_mutator_unit_test.go
Comment (sreyasn) [inline webhooks/virtualmachine/mutation/virtualmachine_mutator_unit_test.go:306]: can we add another test where ctx.vm.Spec.BiosUUID and  vm.Spec.Bootstrap.CloudInit.InstanceId is preset?.

Comment (dougm) [reply webhooks/virtualmachine/mutation/virtualmachine_mutator_unit_test.go:306]: Added

File: api/v1alpha3/virtualmachine_bootstrap_types.go
Comment (sreyasn) [inline api/v1alpha3/virtualmachine_bootstrap_types.go:92]: Can both devOps and Admin users specify this field ?

**Review by sreyasn — APPROVED** [review]
The change looks good to me!. Thanks @dougm.

File: api/v1alpha3/virtualmachine_bootstrap_types.go
Comment (dougm) [reply api/v1alpha3/virtualmachine_bootstrap_types.go:92]: Currently, yes, I don't think we called out a need for this field to be admin-only.

**Pull request merged**
